### ISSUE TITLE
How to use label, to organize key value pairs like webdatasets?

### ISSUE THREAD

**Issue by yangahau** [awslabs/s3-connector-for-pytorch#310]
### Tell us more about this new feature.

My department's previous code used webdatasets，like as:
``` python
#dataset.tar
#├── 00001.jpg
#├── 00001.cls
#├── 00002.jpg
#├── 00002.cls
#├── ...
dataset_path = "dataset.tar"

dataset = wds.WebDataset(dataset_path).decode("rgb").to_tuple("jpg", "cls")
```

If I want to use s3torchconnector, how can I implement labeled return values in the dataloader？

**Comment by IsaevIlya** [comment]
Hello @yangahau,
Thank you for your interest in s3-connector-for-pytorch. Our library can indeed provide streamed data from S3 to WebDataset.Iterators. However, you'll need to handle the output manually. Below, I've provided an example of how you can create an LabeledS3Dataset on top of S3IterableDataset and WebDataset.Iterators.

Please note that this is not production-ready code; it's merely an example to demonstrate how you can handle the combination of an S3 dataset and WebDataset:

```python
import s3torchconnector
import webdataset as wbs
import torch
import io
from typing import Optional, Tuple, List, Dict, Any, Iterator
from PIL import Image
import torchvision.transforms as transforms
from dataclasses import dataclass
from torch.utils.data import IterableDataset, DataLoader
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


@dataclass
class Sample:
    """Sample data structure"""

    name: str
    data: Any
    is_image: bool


def process_sample(sample: Dict[str, Any]) -> Optional[Sample]:
    """Extract and validate data from a sample"""
    try:
        if "fname" not in sample or "data" not in sample:
            return None

        is_img = sample["fname"].endswith(".jpg")
        name = sample["fname"].rsplit(".", 1)[0]

        if is_img:
            data = Image.open(io.BytesIO(sample["data"]))
            data.verify()  # Validate image
            data = Image.open(io.BytesIO(sample["data"]))
        else:
            # Parse labeling info according to your schema
            # In this sample lables placed on separate lines
            data = sample["data"].decode().strip().split("\n")

        return Sample(name, data, is_img)
    except Exception as e:
        logger.warning(f"Error processing sample: {e}")
        return None


class LabeledS3Dataset(IterableDataset):
    """Dataset for loading labeled image data from S3"""

    def __init__(
        self,
        shards_uri: str,
        region: str,
        config: Optional[s3torchconnector.S3ClientConfig] = None,
        transform: Optional[transforms.Compose] = None,
    ):
        super().__init__()
        self.transform = transform or transforms.ToTensor()

        # Initialize S3 dataset
        self.s3_dataset = s3torchconnector.S3IterableDataset.from_prefix(
            shards_uri,
            region=region,
            s3client_config=config,
            transform=lambda obj: {"url": obj.key, "stream": obj},
        )

        # Temporary storage for matching images with cls files
        self.items: Dict[str, Any] = {}

    def __iter__(self) -> Iterator[Tuple[str, torch.Tensor, List[str]]]:
        """Iterate over dataset, yielding (name, image, label) tuples"""
        for sample in wbs.tariterators.tar_file_expander(self.s3_dataset):
            try:
                sample_data = process_sample(sample)
                if not sample_data:
                    continue

                if sample_data.name in self.items:
                    # Match found - yield pair and cleanup
                    if sample_data.is_image:
                        image, label = sample_data.data, self.items[sample_data.name]
                    else:
                        image, label = self.items[sample_data.name], sample_data.data
                    logger.info(f"Match found for {sample_data.name}")
                    # Cleanup record from dictionary to prevent memory pollution
                    del self.items[sample_data.name]

                    try:
                        image_tensor = self.transform(image)
                        yield sample_data.name, image_tensor, label
                    except Exception as e:
                        logger.warning(f"Transform error for {sample_data.name}: {e}")
                else:
                    # Store for later matching
                    self.items[sample_data.name] = sample_data.data

            except Exception as e:
                logger.error(f"Iterator error: {e}")
                continue
```

To use this dataset:
```python
REGION = "us-east-1"
SHARDS_URI = "XXXXXXXXXXXX"

# Usage example:
dataset = LabeledS3Dataset(
    SHARDS_URI,
    region=REGION,
)

# Create DataLoader
dataloader = DataLoader(dataset)

# Test the dataset
for name, image, label in dataloader:
    print(f"Sample name: {name}")
    print(f"Sample labels: {label}")
    print(f"Image size: {image.shape}")
```
Sample output:
```bash
INFO:s3torchconnector.s3iterable_dataset:Building S3IterableDataset from_prefix s3_uri='XXXXXXXXXXXX'
INFO:__main__:Match found for 00000
Sample name: ('00000',)
Sample labels: [('1',), ('88',)]
Image size: torch.Size([1, 4, 1489, 819])
INFO:__main__:Match found for 00001
Sample name: ('00001',)
Sample labels: [('12',), ('34',), ('23',)]
Image size: torch.Size([1, 4, 464, 1056])
INFO:__main__:Match found for 00002
Sample name: ('00002',)
Sample labels: [('31',), ('67',)]
Image size: torch.Size([1, 4, 464, 1023])
INFO:__main__:Match found for 00003
Sample name: ('00003',)
Sample labels: [('44',), ('76',), ('99',), ('101',)]
Image size: torch.Size([1, 4, 447, 853])
INFO:__main__:Match found for 00004
Sample name: ('00004',)
Sample labels: [('1',)]
Image size: torch.Size([1, 4, 464, 1012])
```

This implementation allows you to stream data from S3, process it using WebDataset, and yield it as PyTorch tensors. Remember to adjust the code according to your specific data structure and requirements.

If you have any questions about this implementation or need further clarification, please don't hesitate to ask.

**Comment by yangahau** [comment]
@IsaevIlya 
Thank you for your detailed answer. 
But what I want to know more is how to read sample data and labels only on s3torchconnector. 
In the above code example, if s3torchconnector is only used for reading tar files, why don't I use webdataset for all of it without s3torchconnector.

**Comment by IsaevIlya** [comment]
Hey @yangahau,
The s3-torch-connector library focuses on providing an efficient transport layer for S3. If you need to read WebDataset files from S3 with high performance, you can use our library as the transport layer and compare its speed with WebDataset's built-in transport. Currently, we are not planning to develop our own WebDataset implementation.
